google/cloud/forseti/notifier/notifier.py

```diff
@@ -136,70 +138,72 @@ def run(inventory_index_id, progress_queue, service_config=None):
-
-    violations = scanner_dao.map_by_resource(violations_as_dict)
-
-    for retrieved_v in violations:
-        log_message = ('Retrieved {} violations for resource \'{}\''.format(
-            len(violations[retrieved_v]), retrieved_v))
-        LOGGER.info(log_message)
+        scanner_index_id = scanner_dao.get_latest_scanner_index_id(
+            session, inventory_index_id)
+        if not scanner_index_id:
+            LOGGER.error(
+                'No success or partial success scanner index found for '
+                'inventory index: "%s".', inventory_index_id)
+        else:
+            # get violations
+            violation_access = scanner_dao.ViolationAccess(session)
+            violations = violation_access.list(
+                scanner_index_id=scanner_index_id)
+            violations = convert_to_timestamp(session, violations)
+            violations_as_dict = []
+            for violation in violations:
+                violations_as_dict.append(
+                    scanner_dao.convert_sqlalchemy_object_to_dict(violation))
+            violation_map = scanner_dao.map_by_resource(violations_as_dict)
+
+            for retrieved_v in violation_map:
+                log_message = (
+                    'Retrieved {} violations for resource \'{}\''.format(
+                        len(violation_map[retrieved_v]), retrieved_v))
+                LOGGER.info(log_message)
+                progress_queue.put(log_message)
+
+            # build notification notifiers
+            notifiers = []
+            for resource in notifier_configs['resources']:
+                if violation_map.get(resource['resource']) is None:
+                    log_message = 'Resource \'{}\' has no violations'.format(
+                        resource['resource'])
+                    progress_queue.put(log_message)
+                    LOGGER.info(log_message)
+                    continue
+                if not resource['should_notify']:
+                    LOGGER.debug('Not notifying for: %s', resource['resource'])
+                    continue
+                for notifier in resource['notifiers']:
+                    log_message = (
+                        'Running \'{}\' notifier for resource \'{}\''.format(
+                            notifier['name'], resource['resource']))
+                    progress_queue.put(log_message)
+                    LOGGER.info(log_message)
+                    chosen_pipeline = find_notifiers(notifier['name'])
+                    notifiers.append(chosen_pipeline(
+                        resource['resource'], inventory_index_id,
+                        violation_map[resource['resource']], global_configs,
+                        notifier_configs, notifier['configuration']))
+
+            # Run the notifiers.
+            for notifier in notifiers:
+                notifier.run()
+
+        if (notifier_configs.get('violation') and
```

Comment:
This should be moved under the `else` block.  i.e. we don't want this to run if there is no valid scanner index id.

Reply:
Done.